base-images-sbom-script/app/base_images_sbom_script.py

```diff
@@ -0,0 +1,136 @@
+    """
+    This function expects that the image is in the expected format
+    as generated from the output of
+    "buildah images --format '{{ .Name }}:{{ .Tag }}@{{ .Digest }}'"
+
+    :param image: (str) image reference
+    :return: ParsedImage (namedTuple): the image parsed into individual parts
+    """
+
+    # example image: registry.access.redhat.com/ubi8/ubi:latest@sha256:627867e53ad6846afba2dfbf5cef1d54c868a9025633ef0afd546278d4654eac # noqa
+    # repository_with_tag = registry.access.redhat.com/ubi8/ubi:latest
+    # digest = sha256:627867e53ad6846afba2dfbf5cef1d54c868a9025633ef0afd546278d4654eac
+    # repository_without_tag = registry.access.redhat.com/ubi8/ubi
+    # tag = latest
+    # name = ubi
+    repository_with_tag, digest = image.split("@")
+    # splitting from the right side once on colon, as the repository part
+    # might contain registry url containing a port (host:port)
+    repository_without_tag, tag = repository_with_tag.rsplit(":", 1)
+    if tag == "<none>":
+        tag = None
+    # name is the last fragment of the repository
+    name = repository_without_tag.split("/")[-1]
+
+    return ParsedImage(repository_without_tag=repository_without_tag, digest=digest, tag=tag, name=name)
+
+
+def get_base_images_sbom_components(base_images_digests, is_last_from_scratch):
+    """
+    Creates the base images sbom data
+
+    :param base_images_digests: (List) - list of base images digests, same as BASE_IMAGE_DIGESTS tekton result
+    :param is_last_from_scratch: (Boolean) - Is the last stage/base image from scratch?
+    :return: components (List) - List of dict items in which each item contains sbom data about each base image
+    """
+
+    components = []
+
+    # property_name shows whether the image was used only in the building process
+    # or if it is the final base image. If the final base image is scratch, then
+    # this is omitted, because we aren't including scratch in the sbom.
+    for index, image in enumerate(base_images_digests):
+        property_name = "konflux:container:is_builder_image:for_stage"
+        property_value = str(index)
+        if index == len(base_images_digests) - 1 and not is_last_from_scratch:
+            property_name = "konflux:container:is_base_image"
+            property_value = "true"
+
+        parsed_image = parse_image_reference_to_parts(image)
+
+        purl = PackageURL(
+            type="oci",
+            name=parsed_image.name,
+            version=parsed_image.digest,
+            qualifiers={
+                "repository_url": parsed_image.repository_without_tag,
+                "tag": parsed_image.tag,
+            },
+        )
+
+        component = {
+            "type": "container",
+            "name": parsed_image.repository_without_tag,
+            "purl": purl.to_string(),
+            "properties": [{"name": property_name, "value": property_value}],
+        }
+        components.append(component)
```

Comment:
Technically, it's possible to use the same image (same `repository_without_tag` and `digest`) in more than one build stage, or to use it both in a build and in the last stage.

In such a case, IMO we should report the image only once and combine its properties, e.g.

```json
{
  "type": "container",
  "name": "registry.access.redhat.com/ubi8/ubi",
  "purl": "pkg:oci/ubi@sha256:627867e53ad6846afba2dfbf5cef1d54c868a9025633ef0afd546278d4654eac?repository_url=registry.access.redhat.com/ubi8/ubi&tag=latest",
  "properties": [
    {
      "name": "konflux:container:is_builder_image:for_stage",
      "value": "0"
    },
    {
      "name": "konflux:container:is_base_image",
      "value": "true"
    }
  ]
}
```

Reporting the same component twice would not be correct because AFAIK purls must be unique.

Reply:
And now that I think about it, maybe we don't need to include the `tag` in the purl - otherwise we'd have to think about what to do if the same image appears twice but with different tags

Reply:
Removed the tags and added code for handling reuse of base images together with tests and I mentioned it in the README as well